File: cpp_100-999/931.cc
// C++
// leetcode 931
// https://leetcode-cn.com/problems/minimum-falling-path-sum/
// 


class Solution {
public:
    int min3(int a, int b, int c) {
        if(b < a)
            a = b;
        if(c < a)
            a = c;
        return  a;
    }

    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for(int i = 1; i < n; i++) {
            for(int j = 0; j < n; j++) {
                int a = j - 1;
                int b = j;
                int c = j + 1;
                if(a < 0)
                    a++;
                if(c == n)
                    c--;
                matrix[i][j] += min3(matrix[i-1][a], matrix[i-1][b], matrix[i-1][c]);
            }
        }
        
        int res = INT_MAX;
        for(auto x : matrix[n - 1])
            res = min(res, x);
        return  res;
    }
};

File: Go/Go_100-999/931.go
// Go
// leetcode 931
// https://leetcode.cn/problems/minimum-falling-path-sum/

func minFallingPathSum(matrix [][]int) int {
    m, n := len(matrix), len(matrix[0])
    for i:= m-2; i >= 0; i-- {
        for j := 0; j<n; j++ {
            cur := matrix[i+1][j]
			if j >0 {
				cur = min(cur, matrix[i+1][j-1])
			}

			if j <n-1 {
				cur = min(cur, matrix[i+1][j+1])
			}
			matrix[i][j] += cur
        }
    }
	res := 100_000_000
	for _, x := range matrix[0] {
		res = min(res, x)
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}